### DISCUSSION TITLE
USB Device Interface Capabilities

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#36070] [Q&A] [answered]
My understanding is that Zephyr's USB support is considered USB "device" capabilities rather than USB "host" capabilities.

I need to interface with a USB 2.0 specification USB device.

Does the Zephyr stack support reading and writing to such a USB device, or is this simply not possible and a show stopper?

Is the distinction between USB host and device significant for a USB 2.0 device?

Assuming it has to be a USB host to interface with an external USB device, has there been any effort to support that in Zephyr or 3rd party add-ons?

**Comment by nashif** [comment]
@jfischer-no FYI

**Comment by jfischer-no** [comment, marked as answer]
> My understanding is that Zephyr's USB support is considered USB "device" capabilities rather than USB "host" capabilities.
> 

In general, there is no strategic preference.

> I need to interface with a USB 2.0 specification USB device.
> 
> Does the Zephyr stack support reading and writing to such a USB device, or is this simply not possible and a show stopper?

Zephyr does not yet have support for USB host, so if "USB device" that you mean, is an USB keyboard for example, than it is not possible.

> 
> Is the distinction between USB host and device significant for a USB 2.0 device?

Yes

> 
> Assuming it has to be a USB host to interface with an external USB device, has there been any effort to support that in Zephyr or 3rd party add-ons?

yes, see https://github.com/zephyrproject-rtos/zephyr/pull/30361

**Reply by bob2oneil** [reply]
Do you know off hand which branch pull will have this USB Host Experimental drivers?

**Reply by bob2oneil** [reply]
I have the branch and source code reviewed.  Can you confirm that support for USB host capabilities requires a hardware chipset.  That is what is in place for this effort, meaning it is a driver for a specific chipset(s) and boards (2).

**Reply by jfischer-no** [reply]
> I have the branch and source code reviewed. Can you confirm that support for USB host capabilities requires a hardware chipset.

Yes, you need a SoC with integrated USB host controller or an external USB host controller.

>That is what is in place for this effort, meaning it is a driver for a specific chipset(s) and boards (2).

Do not get it, is it a question?

Are you kidding me?

**Reply by bob2oneil** [reply]
Bad language or misunderstanding.  What I was referring to was the USB Host effort that is posted as a potential branch to merge into the repo. I was done apparently by a graduate student in Germany, possibly associated with you?  In that effort, there were 2 separate board targets, both of which had in their device trees a USBH element, suggesting that there needed to be a USB Host chipset on the board to get this behavior.  In the case of both of these boards, that was true.  Another way of stating this as you have confirmed, to get USB Host behavior, currently there is no pure software solution without some hardware support.  Thanks for the confirmation.

**Reply by bob2oneil** [reply]
Johann, we tried to get this pull request in its entire working on an  NXP1050 board, and it bricked the board.  Do you have any insight into this development in terms of its working status, or current history?  What I really want to know is if it ever worked, because that way I can see if a mistake was made on our part that needs more investigation.

Thanks in advance.
